core/server/api_container/server/service_network/default_service_network.go

```diff
@@ -1791,3 +1878,102 @@ func renderTemplateToFile(templateAsAString string, templateData interface{}, de
 	}
 	return nil
 }
+
+func waitUntilAllTCPAndUDPPortsAreOpen(
+	ipAddr net.IP,
+	ports map[string]*port_spec.PortSpec,
+) error {
+
+	var portCheckErrorGroup errgroup.Group
+
+	for _, portSpec := range ports {
+		wrappedWaitFunc := func() error {
+			return waitUntilPortIsOpenWithTimeout(ipAddr, portSpec)
+		}
+		portCheckErrorGroup.Go(wrappedWaitFunc)
+	}
+	//This error group pattern allow us to reject early if at least one of the ports check fails
+	//in this opportunity we want to reject early because the main user pain that we want to handle
+	//is a wrong service configuration
+	if err := portCheckErrorGroup.Wait(); err != nil {
+		return stacktrace.Propagate(err, "An error occurred while waiting for all TCP and UDP ports to be open")
+	}
+
+	return nil
+}
+
+func waitUntilPortIsOpenWithTimeout(
+	ipAddr net.IP,
+	portSpec *port_spec.PortSpec,
+) error {
+	// reject early if it's disable
+	if !portSpec.GetWait().GetEnable() {
+		return nil
+	}
+	timeout := portSpec.GetWait().GetTimeout()
+
+	var (
+		ticker                  = time.NewTicker(timeout)
+		startTime               = time.Now()
+		finishTime              = startTime.Add(timeout)
+		shouldContinueInTheLoop = true
+		retries                 = 0
+		err                     error
+	)
+
+	time.Sleep(portSpec.GetWait().GetInitialDelay())
```

Comment:
Interesting... So you're computing the finishTime BEFORE you do the initial delay. Can it lead to weird situation where, if I do `initialDelay=60*time.Second, timeout=30*time.Second` it actually throws without trying at least once?
I find this way of thinking about timeout VS initialDelay a bit confusing. To me, it's more natural to start accounting for the timeout once the initial delay has expired. 
1. Sleep for the initial delay
2. Compute finishTime from now
3. Start the retries until timeout has been reached or port is open, whichever comes first

this also has the advantage of completely decoupling the initial delay and timeout, having a less risky logic

Reply:
I removed the initial delay, it's not something that will provide now, v0 will include only timeouts, more here: https://www.notion.so/kurtosistech/Service-Ports-Wait-Design-Doc-25cc5ea9ab6f4d48a2f2917e4f25c5f5?pvs=4#ce1d2fdf6e5a4ca4a56c3a5995341523